### PR TITLE
[SPARK-35879][Core][Shuffle] Fix performance regression caused by collectFetchRequests

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

This PR fixes perf regression at the executor side when creating fetch requests with large initial partitions

image

In NetEase, we had an online job that took `45min` to "fetch" about 100MB of shuffle data, which actually turned out that it was just collecting fetch requests slowly. Normally, such a task should finish in seconds.

See the `DEBUG` log

```
21/06/22 11:52:26 DEBUG BlockManagerStorageEndpoint: Sent response: 0 to kyuubi.163.org:
21/06/22 11:53:05 DEBUG ShuffleBlockFetcherIterator: Creating fetch request of 3941440 at BlockManagerId(12, .., 43559, None) with 19 blocks
21/06/22 11:53:44 DEBUG ShuffleBlockFetcherIterator: Creating fetch request of 3716400 at BlockManagerId(20, .., 38287, None) with 18 blocks
21/06/22 11:54:41 DEBUG ShuffleBlockFetcherIterator: Creating fetch request of 4559280 at BlockManagerId(6, .., 39689, None) with 22 blocks
21/06/22 11:55:08 DEBUG ShuffleBlockFetcherIterator: Creating fetch request of 3120160 at BlockManagerId(33, .., 39449, None) with 15 blocks
```

I also create a test case locally with my local laptop docker env to give some reproducible cases.

```
bin/spark-sql --conf spark.kubernetes.file.upload.path=./ --master k8s://https://kubernetes.docker.internal:6443 --conf spark.kubernetes.container.image=yaooqinn/spark:v20210624-5 -c spark.kubernetes.context=docker-for-desktop_1 --num-executors 5 --driver-memory 5g --conf spark.kubernetes.executor.podNamePrefix=sparksql
```

```sql
 SET spark.sql.adaptive.enabled=true;
 SET spark.sql.shuffle.partitions=3000;
 SELECT /*+ REPARTITION */ 1 as pid, id from range(1, 1000000, 1, 500);
 SELECT /*+ REPARTITION(pid, id) */ 1 as pid, id from range(1, 1000000, 1, 500);
 ```

### Why are the changes needed?

fix perf regression which was introduced by SPARK-29292 (3ad4863673fc46080dda963be3055a3e554cfbc7) in v3.1.0.

3ad4863673fc46080dda963be3055a3e554cfbc7 is for support compilation with scala 2.13 but the performance losses is huge. We need to consider backporting this PR to branch 3.1.

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

Mannully,

#### before
```log
 21/06/23 13:54:22 DEBUG ShuffleBlockFetcherIterator: maxBytesInFlight: 50331648, targetRemoteRequestSize: 10066329, maxBlocksInFlightPerAddress: 2147483647
 21/06/23 13:54:38 DEBUG ShuffleBlockFetcherIterator: Creating fetch request of 2314708 at BlockManagerId(2, 10.1.3.114, 36423, None) with 86 blocks
 21/06/23 13:54:59 DEBUG ShuffleBlockFetcherIterator: Creating fetch request of 2636612 at BlockManagerId(3, 10.1.3.115, 34293, None) with 87 blocks
 21/06/23 13:55:18 DEBUG ShuffleBlockFetcherIterator: Creating fetch request of 2508706 at BlockManagerId(4, 10.1.3.116, 41869, None) with 90 blocks
 21/06/23 13:55:34 DEBUG ShuffleBlockFetcherIterator: Creating fetch request of 2350854 at BlockManagerId(5, 10.1.3.117, 45787, None) with 85 blocks
 21/06/23 13:55:34 INFO ShuffleBlockFetcherIterator: Getting 438 (11.8 MiB) non-empty blocks including 90 (2.5 MiB) local and 0 (0.0 B) host-local and 348 (9.4 MiB) remote blocks
 21/06/23 13:55:34 DEBUG ShuffleBlockFetcherIterator: Sending request for 87 blocks (2.5 MiB) from 10.1.3.115:34293
 21/06/23 13:55:34 INFO TransportClientFactory: Successfully created connection to /10.1.3.115:34293 after 1 ms (0 ms spent in bootstraps)
 21/06/23 13:55:34 DEBUG ShuffleBlockFetcherIterator: Sending request for 90 blocks (2.4 MiB) from 10.1.3.116:41869
 21/06/23 13:55:34 INFO TransportClientFactory: Successfully created connection to /10.1.3.116:41869 after 2 ms (0 ms spent in bootstraps)
 21/06/23 13:55:34 DEBUG ShuffleBlockFetcherIterator: Sending request for 85 blocks (2.2 MiB) from 10.1.3.117:45787
 ```
```log
 21/06/23 14:00:45 INFO MapOutputTracker: Broadcast outputstatuses size = 411, actual size = 828997
 21/06/23 14:00:45 INFO MapOutputTrackerWorker: Got the map output locations
 21/06/23 14:00:45 DEBUG ShuffleBlockFetcherIterator: maxBytesInFlight: 50331648, targetRemoteRequestSize: 10066329, maxBlocksInFlightPerAddress: 2147483647
 21/06/23 14:00:55 DEBUG ShuffleBlockFetcherIterator: Creating fetch request of 1894389 at BlockManagerId(2, 10.1.3.114, 36423, None) with 99 blocks
 21/06/23 14:01:04 DEBUG ShuffleBlockFetcherIterator: Creating fetch request of 1919993 at BlockManagerId(3, 10.1.3.115, 34293, None) with 100 blocks
 21/06/23 14:01:14 DEBUG ShuffleBlockFetcherIterator: Creating fetch request of 1977186 at BlockManagerId(5, 10.1.3.117, 45787, None) with 103 blocks
 21/06/23 14:01:23 DEBUG ShuffleBlockFetcherIterator: Creating fetch request of 1938336 at BlockManagerId(4, 10.1.3.116, 41869, None) with 101 blocks
 21/06/23 14:01:23 INFO ShuffleBlockFetcherIterator: Getting 500 (9.1 MiB) non-empty blocks including 97 (1820.3 KiB) local and 0 (0.0 B) host-local and 403 (7.4 MiB) remote blocks
 21/06/23 14:01:23 DEBUG ShuffleBlockFetcherIterator: Sending request for 101 blocks (1892.9 KiB) from 10.1.3.116:41869
 21/06/23 14:01:23 DEBUG ShuffleBlockFetcherIterator: Sending request for 103 blocks (1930.8 KiB) from 10.1.3.117:45787
 21/06/23 14:01:23 DEBUG ShuffleBlockFetcherIterator: Sending request for 99 blocks (1850.0 KiB) from 10.1.3.114:36423
 21/06/23 14:01:23 DEBUG ShuffleBlockFetcherIterator: Sending request for 100 blocks (1875.0 KiB) from 10.1.3.115:34293
 21/06/23 14:01:23 INFO ShuffleBlockFetcherIterator: Started 4 remote fetches in 37889 ms
 ```

#### After

```log
21/06/24 13:01:16 DEBUG ShuffleBlockFetcherIterator: maxBytesInFlight: 50331648, targetRemoteRequestSize: 10066329, maxBlocksInFlightPerAddress: 2147483647
21/06/24 13:01:16 INFO ShuffleBlockFetcherIterator: ==> Call blockInfos.map(_._2).sum: 40 ms
21/06/24 13:01:16 INFO ShuffleBlockFetcherIterator: ==> Call mergeFetchBlockInfo for shuffle_0_9_2990_2997/9: 0 ms
21/06/24 13:01:16 INFO ShuffleBlockFetcherIterator: ==> Call mergeFetchBlockInfo for shuffle_0_15_2395_2997/15: 0 ms
```

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc @cloud-fan @maropu @HyukjinKwon @srowen thanks very much

File: core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala [hunk 1]
```diff
@@ -435,26 +438,24 @@ final class ShuffleBlockFetcherIterator(
...
+          collectedRemoteRequests).to[ArrayBuffer]
         curRequestSize = curBlocks.map(_.size).sum
       }
     }
     // Add in the final request
     if (curBlocks.nonEmpty) {
-      curBlocks = createFetchRequests(curBlocks, address, isLast = true,
-        collectedRemoteRequests)
-      curRequestSize = curBlocks.map(_.size).sum
+      createFetchRequests(curBlocks, address, isLast = true, collectedRemoteRequests)
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala:458]: why can we skip updating `curBlocks` and `curRequestSize` here?

Comment (cloud-fan) [reply core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala:458]: ah, we are leaving this method and no need to update local variables here.

File: core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala [hunk 2]
```diff
@@ -435,26 +438,24 @@ final class ShuffleBlockFetcherIterator(
...
       assertPositiveBlockSize(blockId, size)
-      curBlocks = curBlocks ++ Seq(FetchBlockInfo(blockId, size, mapIndex))
+      curBlocks += FetchBlockInfo(blockId, size, mapIndex)
       curRequestSize += size
       // For batch fetch, the actual block in flight should count for merged block.
       val mayExceedsMaxBlocks = !doBatchFetch && curBlocks.size >= maxBlocksInFlightPerAddress
       if (curRequestSize >= targetRemoteRequestSize || mayExceedsMaxBlocks) {
         curBlocks = createFetchRequests(curBlocks, address, isLast = false,
-          collectedRemoteRequests)
+          collectedRemoteRequests).to[ArrayBuffer]
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala:452]: how about
```
val newCurBlocks = createFetchRequests...
curBlocks.clear
curBlocks ++= newCurBlocks
```

Then `curBlocks` can be a val.

Comment (yaooqinn) [reply core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala:452]: good idea

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
retest this please

File: core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala (hunk 2 above)
Comment (yaooqinn) [reply core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala:452]: `curBlocks.clear` will clear all the elements which are still needed

**Review by HyukjinKwon — APPROVED** [review]

**Review by Ngone51 — APPROVED** [review]

**Comment by Ngone51** [comment]
cc @mridulm @otterc

**Comment by otterc** [comment]
Looks good to me as well

File: core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala [hunk 3]
```diff
@@ -433,28 +436,25 @@ final class ShuffleBlockFetcherIterator(
...
+      curBlocks += FetchBlockInfo(blockId, size, mapIndex)
       curRequestSize += size
       // For batch fetch, the actual block in flight should count for merged block.
       val mayExceedsMaxBlocks = !doBatchFetch && curBlocks.size >= maxBlocksInFlightPerAddress
       if (curRequestSize >= targetRemoteRequestSize || mayExceedsMaxBlocks) {
-        curBlocks = createFetchRequests(curBlocks, address, isLast = false,
-          collectedRemoteRequests)
+        val rest = createFetchRequests(curBlocks, address, isLast = false, collectedRemoteRequests)
+        curBlocks.clear()
+        curBlocks ++= rest
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala:451]: Instead of clearing it, create new instance ?

Comment (yaooqinn) [reply core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala:451]: > If code evolves to use curBlocks in createFetchRequests or downstream (as Seq)

I not sure it's a good idea to use an ArrayBuffer as Seq implicitly.

Comment (mridulm) [reply core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala:451]: The PR is already doing this for `mergeContinuousShuffleBlockIdsIfNeeded` ?

Comment (yaooqinn) [reply core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala:451]: > The PR is already doing this for `mergeContinuousShuffleBlockIdsIfNeeded`? 
Not by this PR. `mergeContinuousShuffleBlockIdsIfNeeded` takes both `Seq` for handling local/host-local blocks and `ArrayBuffer` for the remotes. This PR modified it to ensure that it always returns a new copy for both `doBatchFetch` and `!doBatchFetch`. And before this PR,  `!doBatchFetch` branch might potentially result in surprises.

Comment (yaooqinn) [reply core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala:451]: > The PR is already doing this for `mergeContinuousShuffleBlockIdsIfNeeded`? 

Not by this PR. `mergeContinuousShuffleBlockIdsIfNeeded` takes both `Seq` for handling local/host-local blocks and `ArrayBuffer` for the remotes. This PR modified it to ensure that it always returns a new copy for both `doBatchFetch` and `!doBatchFetch`. And before this PR,  `!doBatchFetch` branch might potentially result in surprises.

Comment (yaooqinn) [reply core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala:451]: If you insist, I can revert the last commit https://github.com/apache/spark/pull/33063/commits/f8942209f2a00ef32940cd2805474c117fd89d5e. I do not have a strong opinion on it.

**Review by dongjoon-hyun — APPROVED** [review]
After the last reverting commit, it looks better. 
+1, LGTM.

**Comment by dongjoon-hyun** [comment]
Hi, @mridulm . Could you do the review once more?

**Comment by mridulm** [comment]
Thanks for the changes @yaooqinn and thanks for the ping @dongjoon-hyun :-)
The changes look good - but I want to make sure we still have similar performance benefits as we saw in the PR description.
Will it be possible to rerun the benchmark to ensure no regression @yaooqinn ? Thx !

**Comment by mridulm** [comment]
@dongjoon-hyun I am +1 on the change if the performance is still good based on @yaooqinn's benchmark (will be going on vacation soon, so dont want reviews to be blocked on my accord !). Thx

**Comment by yaooqinn** [comment]
@mridulm @dongjoon-hyun, I re-run the benchmark based on the final commit manually. The debug log below shows the performance regression is gone.
```log
21/06/26 04:04:01 INFO MapOutputTrackerWorker: Got the map output locations
21/06/26 04:04:01 DEBUG ShuffleBlockFetcherIterator: maxBytesInFlight: 50331648, targetRemoteRequestSize: 10066329, maxBlocksInFlightPerAddress: 2147483647
21/06/26 04:04:01 DEBUG ShuffleBlockFetcherIterator: Creating fetch request of 2606046 at BlockManagerId(2, 10.1.5.72, 37767, None) with 88 blocks
21/06/26 04:04:01 DEBUG ShuffleBlockFetcherIterator: Collected remote fetch requests for BlockManagerId(2, 10.1.5.72, 37767, None) in 82 ms
```

**Comment by yaooqinn** [comment]
thanks all! merged to master/3.1

**Comment by dongjoon-hyun** [comment]
Thank you, @yaooqinn and @mridulm !

**Comment by mridulm** [comment]
Thanks @yaooqinn !

**Pull request closed without merging**